src/vmm/src/guest_config/templates/mod.rs

```diff
@@ -66,21 +86,183 @@ pub mod x86_64 {
     /// in the context of a CPUID tree.
     #[allow(unused)]
     #[derive(Debug, Deserialize, Eq, PartialEq)]
-    pub struct CpuidRegisterId {
-        leaf: u32,
-        subleaf: u32,
-        flags: crate::guest_config::cpuid::cpuid_ffi::KvmCpuidFlags,
-        pointer: CpuidRegister,
+    pub struct CpuidRegisterModifier {
+        /// Leaf value.
+        #[serde(deserialize_with = "deserialize_u32_from_str")]
+        pub leaf: u32,
+        /// Sub-Leaf value.
+        #[serde(deserialize_with = "deserialize_u32_from_str")]
+        pub subleaf: u32,
+        /// KVM feature flags for this leaf-subleaf.
+        #[serde(deserialize_with = "deserialize_kvm_cpuid_flags")]
+        pub flags: crate::guest_config::cpuid::cpuid_ffi::KvmCpuidFlags,
+        /// CPUID register to be modified by the bitmask.
+        #[serde(deserialize_with = "deserialize_cpuid_register")]
+        pub register: CpuidRegister,
+        /// Mask to be applied to register's value at the address
+        /// provided.
+        pub mask: BitMask,
     }
 
     /// Wrapper type to containing x86_64 CPU config modifiers.
     #[derive(Debug, Deserialize, Eq, PartialEq)]
     pub struct X86_64CpuTemplate {
         /// Modifiers for CPUID configuration.
-        // pub cpuid_modifiers: Vec<RegisterBitMask>,
+        #[serde(default)]
+        pub cpuid_modifiers: Vec<CpuidRegisterModifier>,
         /// Modifiers for model specific registers.
+        #[serde(default)]
         pub msr_modifiers: Vec<RegisterModifier>,
     }
+
+    impl X86_64CpuTemplate {
+        /// Makes of a CPU template with modifiers
+        /// to be applied on top of the host configuration
+        /// to generate the guest configuration to be used.
+        pub fn apply_template(
+            self,
+            host_config: X86_64CpuConfiguration,
+        ) -> Result<X86_64CpuConfiguration, Error> {
+            let mut guest_msrs_map: HashMap<u64, u64> = HashMap::new();
+            let mut guest_cpuid_map: BTreeMap<CpuidKey, CpuidEntry>;
+
+            // Get the hash map of CPUID data
+            if host_config.cpuid.amd().is_some() {
+                guest_cpuid_map = host_config.cpuid.amd().unwrap().0.clone();
+            } else if host_config.cpuid.intel().is_some() {
+                guest_cpuid_map = host_config.cpuid.intel().unwrap().0.clone();
+            } else {
+                return Err(Error::Cpuid);
+            }
+
+            // Apply CPUID modifiers
+            for modifier in self.cpuid_modifiers {
```

Comment:
I think it might be a good idea to extract this and the other loop into separate functions.

Reply:
I plan to pull out the MSRs loop into its own function as the same logic will be applicable for aarch config.